plutus-contract/src/Plutus/Contract/Test/ContractModel.hs

```diff
@@ -1134,3 +1149,79 @@ checkNoLockedFundsProof options spec NoLockedFundsProof{nlfpMainStrategy   = mai
                               "  " ++ show bal'
                     DL.assert err (bal `leq` bal')
 
+-- | A whitelist entry tells you what final log entry prefixes
+-- are acceptable for a given error
+data WhitelistEntry = WhitelistEntry { acceptEmptyLog :: Bool, errorPrefixes :: [Text.Text] }
+-- | A whitelist associates errors with a whitelist entry
+type Whitelist = Map String WhitelistEntry
+
+-- | Check that a log is accepted by a whitelist entry
+isAcceptedBy :: Maybe Text.Text -> WhitelistEntry -> Bool
+isAcceptedBy Nothing wle          = acceptEmptyLog wle
+isAcceptedBy (Just lastEntry) wle = any (`Text.isPrefixOf` lastEntry) (errorPrefixes wle)
+
+{- Note [Maintaining `whitelistOk` and `checkErrorWhitelist`]
+   The intended use case of `checkErrorWhitelist` is to be able to assert that failures of
+   validation only happen for reasons that the programmer intended. However, to avoid
+   degenerate whitelists that accept obvious mistakes like validation failing because a
+   partial function in the prelude failed we need to make sure that any whitelist used passes
+   the `whitelistOk` check. This means in turn that we need to maintain `whitelistOk` when we
+   introduce new failure modes or change existing failure modes in the prelude or elsewhere
+   in the plutus system.
+-}
+
+{- Note [Divide by zero]
+   Errors that arise from a division by zero fail silently with a `CekEvaluationFailure`. This
+   means that we can not be sure to catch this failure using our approach if the programmer
+   has whitelisted some log message that happens to be the last message in the log. This is
+   a big problem that I (Max) don't know how to get around at the time of writing. The best
+   fix would be to fix the CEK machine implementaiton to throw a different type of error when
+   division by zero fails.
+-}
+
+-- | Check that a whitelist does not accept any partial functions
+whitelistOk :: Whitelist -> Bool
+whitelistOk wl = noPreludePartials
+  where
+    noPreludePartials = case Map.lookup "CekEvaluationFailure" wl of
+      -- We specifically ignore `checkHasFailed` here because it is the failure you get when a
+      -- validator that returns a boolean fails correctly.
+      Just wle -> all (Prelude.not . (`isAcceptedBy` wle) . Just . Builtins.fromBuiltin) (map fst allErrorCodes \\ [checkHasFailedError])
+               && Prelude.not (Nothing `isAcceptedBy` wle) -- Covers the case for divide by zero with an empty log
+      Nothing  -> True
+
+emptyWhitelist :: Whitelist
+emptyWhitelist = Map.singleton "CekEvaluationFailure" (WhitelistEntry False [Builtins.fromBuiltin checkHasFailedError])
```

Comment:
why not the empty map? overall I'm not sure I get why you have this outer layer of map...

Reply:
Right, I want to hedge against other failure error messages than `"CekEvaluationFailure"` being used / useful in the future / now. However, if there are no other reasonable failure strings than `"CekEvaluationFailure"` then this can safely be flattened.

Reply:
That is, it can safely be flattened to a list of prefixes and nothing else.

Reply:
Oh, and the reason for not having the empty map here is that the name of the whitelist is really wrong. It should probably be `defaultWhitelist`.